src/cargo/util/errors.rs

```diff
@@ -8,144 +7,100 @@ use std::str;
+        CargoError, CargoErrorKind, CargoResultExt, CargoResult;
     }
-}
 
-impl<T, E: CargoError + 'static> ChainError<T> for Result<T, E> {
-    fn chain_error<E2: 'static, C>(self, callback: C) -> CargoResult<T>
-                         where E2: CargoError, C: FnOnce() -> E2 {
-        self.map_err(move |err| {
-            Box::new(ChainedError {
-                error: callback(),
-                cause: Box::new(err),
-            }) as Box<CargoError>
-        })
+    links {
+        Network(NetworkError, NetworkErrorKind);
     }
-}
 
-impl<T> ChainError<T> for Box<CargoError> {
-    fn chain_error<E2, C>(self, callback: C) -> CargoResult<T>
-                         where E2: CargoError, C: FnOnce() -> E2 {
-        Err(Box::new(ChainedError {
-            error: callback(),
-            cause: self,
-        }))
+    foreign_links {
+        ParseSemver(semver::ReqParseError);
+        Io(io::Error);
+        SerdeJson(serde_json::Error);
+        TomlSer(toml::ser::Error);
+        TomlDe(toml::de::Error);
+        Term(term::Error);
+        ParseInt(num::ParseIntError);
+        ParseBool(str::ParseBoolError);
+        Parse(string::ParseError);
+        Git(git2::Error);
+        Curl(curl::Error);
     }
-}
 
-impl<T> ChainError<T> for Option<T> {
-    fn chain_error<E: 'static, C>(self, callback: C) -> CargoResult<T>
-                         where E: CargoError, C: FnOnce() -> E {
-        match self {
-            Some(t) => Ok(t),
-            None => Err(Box::new(callback())),
+    errors {
+        Internal(description: String, details: Option<String>){
+            description(&description)
+            display("{}", &description)
+        }
+        ProcessErrorKind(proc_err: ProcessError) {
+            description(&proc_err.desc)
+            display("{}", &proc_err.desc)
+        }
+        CargoTestErrorKind(test_err: CargoTestError) {
+            description(&test_err.desc)
+            display("{}", &test_err.desc)
         }
     }
 }
 
-impl<E: Error> Error for ChainedError<E> {
-    fn description(&self) -> &str { self.error.description() }
-}
+impl CargoError {
+    pub fn to_internal(self) -> Self {
+        CargoError(CargoErrorKind::Internal(self.description().to_string(), None), self.1)
+    }
 
-impl<E: fmt::Display> fmt::Display for ChainedError<E> {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        fmt::Display::fmt(&self.error, f)
+    fn is_human(&self) -> bool {
+        match &self.0 {
+            &CargoErrorKind::Msg(_) => true,
+            &CargoErrorKind::TomlSer(_) => true,
+            &CargoErrorKind::TomlDe(_) => true,
+            &CargoErrorKind::Curl(_) => true,
+            &CargoErrorKind::Network(ref net_err) => {
+                match net_err {
+                    &NetworkErrorKind::HttpNot200(_, _) => true,
+                    &NetworkErrorKind::Curl(_) => true,
+                    _ => false
+                }
+            },
+            _ => false
```

Comment:
Just to make sure we explicitly list it out, mind expanding out this `_` to match everything?